File: src/main/java/net/swamphut/swampium/core/swobject/instance/SwampiumInstanceManager.java
package net.swamphut.swampium.core.swobject.instance;

import net.swamphut.swampium.core.exception.ServiceInstantiateException;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;

public class SwampiumInstanceManager implements InstanceManager {
    private HashMap<Class<?>, Object> classInstanceMap = new HashMap<>();

    @Override
    public <T> T getInstance(Class<T> clazz) {
        if (!classInstanceMap.containsKey(clazz)) {
            classInstanceMap.put(clazz, instantiateService(clazz));
        }
        return (T) classInstanceMap.get(clazz);
    }

    @Override
    public void removeInstance(Class clazz) {
        if (!classInstanceMap.containsKey(clazz)) throw new IllegalStateException();
        classInstanceMap.remove(clazz);
    }

    private <T> T instantiateService(Class<T> clazz) {
        try {
            return clazz.getConstructor().newInstance();
        } catch (IllegalAccessException | NoSuchMethodException e) {
            throw new ServiceInstantiateException(
                    "SwObject classes are required to have a public default constructor", e, clazz);
        } catch (InvocationTargetException | InstantiationException e) {
            throw new ServiceInstantiateException("Exception throwed while instantiate", e, clazz);
        }
    }
}

File: src/main/java/net/swamphut/swampium/core/swobject/instance/SwampiumInstanceManager.kt
package net.swamphut.swampium.core.swobject.instance

import java.util.*
import kotlin.reflect.KClass


class SwampiumInstanceManager : SwObjectInstanceManager {
    private val instanceMap = HashMap<KClass<out Any>, Any>();
    override fun destroyInstance(instance: Any) {
        instanceMap.remove(instance::class)
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> getInstance(swObjectClass: KClass<T>): T? {
        return instanceMap[swObjectClass] as T?
    }

    override fun putInstance(instance: Any) {
        instanceMap[instance::class] = instance
    }

    /**
     * This function enable swampium core to access basic injectable object before injection services are ready
     */
    internal fun <T : Any> getOrConstructWithoutInjection(swObjectClass: KClass<T>): T {
        getInstance(swObjectClass)?.let { return it }
        if (swObjectClass.constructors.size == 1 && swObjectClass.constructors.first().parameters.isNotEmpty())
            throw IllegalStateException();
        swObjectClass.constructors.first().call().also {
            putInstance(it);
            return it;
        }
    }

}
